route-rs-packets/src/ipv4.rs

```diff
@@ -0,0 +1,344 @@
+        self.validated_checksum = false;
+    }
+
+    pub fn checksum(&self) -> u16 {
+        u16::from_be_bytes(
+            self.data[self.packet_offset + 10..=self.packet_offset + 11]
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    pub fn dcsp(&self) -> u8 {
+        self.data[self.packet_offset + 1] >> 2
+    }
+
+    pub fn ecn(&self) -> u8 {
+        self.data[self.packet_offset + 1] & 0x03
+    }
+
+    pub fn indentification(&self) -> u16 {
+        u16::from_be_bytes(
+            self.data[self.packet_offset + 4..=self.packet_offset + 5]
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    pub fn fragment_offet(&self) -> u16 {
+        u16::from_be_bytes([
+            self.data[self.packet_offset + 6] & 0x1F,
+            self.data[self.packet_offset + 7],
+        ])
+    }
+
+    /// Returns tuple of (Don't Fragment, More Fragments)
+    pub fn flags(&self) -> (bool, bool) {
+        let df = (self.data[self.packet_offset + 6] & 0x40) != 0;
+        let mf = (self.data[self.packet_offset + 6] & 0x20) != 0;
+        (df, mf)
+    }
+
+    /// Verifies the IP header checksum, returns the value and also sets
+    /// the internal bookeeping field. As such we need a mutable reference.
+    pub fn validate_checksum(&mut self) -> bool {
+        let full_sum = &self.data[self.packet_offset..self.payload_offset]
+            .chunks_exact(2)
+            .fold(0, |acc: u32, x| {
+                acc + u32::from(u16::from_be_bytes([x[0], x[1]]))
+            });
+        let (carry, mut sum) = (((full_sum & 0xFFFF_0000) >> 16), (full_sum & 0x0000_FFFF));
+        sum += carry;
+        self.validated_checksum = 0 == (!sum & 0xFFFF);
+        self.validated_checksum
+    }
+
+    /// Calculates what the checksum should be set to given the current header
+    pub fn caclulate_checksum(&self) -> u16 {
+        let full_sum = &self.data[self.packet_offset..self.payload_offset]
+            .chunks_exact(2)
+            .enumerate()
+            .filter(|x| x.0 != 5)
+            .fold(0, |acc: u32, x| {
+                acc + u32::from(u16::from_be_bytes([x.1[0], x.1[1]]))
+            });
+        let (carry, mut sum) = (((full_sum & 0xFFFF_0000) >> 16), (full_sum & 0x0000_FFFF));
+        sum += carry;
+        if sum & 0xFFFF_0000 != 0 {
+            sum += 1;
+        }
+        sum = !sum & 0xFFFF;
+        sum as u16
+    }
+
+    /// Sets checksum field to valid value
+    pub fn set_checksum(&mut self) {
+        let new_checksum = self.caclulate_checksum();
+        self.data[self.packet_offset + 10] = ((new_checksum & 0xFF00) >> 8) as u8;
+        self.data[self.packet_offset + 11] = (new_checksum & 0x00FF) as u8;
+        self.validated_checksum = true;
+    }
+}
+
+/// Returns Ipv4 payload type, reads the header information to get the type
+/// of IpProtocol payload is included. Upon error, returns IpProtocol::Reserved.
```

Comment:
Why not an error type? What if this reserved value is later assigned?

Reply:
Idk making an error type is kind of annoying, this is quick and dirty. Error type is on the list of improvements, i can spawn off an issue.

Reply:
I can fix this by returning a Result at least to start.

Reply:
What kind of errors are possible?

Reply:
There are a few numbers that are undefined?

Reply:
I suppose the error for this is pointless. Nothing could be an Ipv4 packet if it isn't long enough to contain a full header, nor could its version field be incorrect.

So I'll remove the offending check.